sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/CosmosWriterBase.scala

```diff
@@ -91,27 +95,69 @@ private abstract class CosmosWriterBase(
     }
 
     val partitionKeyValue = PartitionKeyHelper.getPartitionKeyPath(objectNode, partitionKeyDefinition)
-    writer.scheduleWrite(partitionKeyValue, objectNode)
+    writer.get.scheduleWrite(partitionKeyValue, objectNode)
   }
 
   override def commit(): WriterCommitMessage = {
     log.logInfo("commit invoked!!!")
-    writer.flushAndClose()
+
+    try {
+      writer.get.flushAndClose()
+    } catch {
+      case bulkWriterStaleError: BulkWriterNoProgressException =>
+        bulkWriterStaleError.activeBulkWriteOperations match {
+          case Some(remainingWriteOperations) =>
+            log.logWarning(s"Error indicating stuck writer when committing writes. Retry will be attempted for "
+              + s"the outstanding ${remainingWriteOperations.size} write operations.", bulkWriterStaleError)
+
+            val bulkWriterForRetry =
+                new BulkWriter(
+                  container,
+                  partitionKeyDefinition,
+                  cosmosWriteConfig,
+                  diagnosticsConfig,
+                  getOutputMetricsPublisher(),
+                  commitAttempt.getAndIncrement())
+            val oldBulkWriter = writer.getAndSet(bulkWriterForRetry)
+
+            cosmosWriteConfig.retryCommitInterceptor match {
+              case Some(onRetryCommitInterceptor) =>
+                log.logInfo("Invoking custom on-retry-commit interceptor...")
+                onRetryCommitInterceptor.beforeRetryCommit()
+              case None =>
+            }
+
+            for (operation <- remainingWriteOperations) {
+              bulkWriterForRetry.scheduleWrite(operation.getPartitionKeyValue, operation.getItem[ObjectNode])
+            }
+            oldBulkWriter.abort(false)
+            bulkWriterForRetry.flushAndClose()
+          // None means not just write operations but also read-many are outstanding we can't retry
+          case None =>
+            log.logError(s"Error indicating stuck writer when committing writes. No retry possible because "
+              + "of outstanding read-many operations.", bulkWriterStaleError)
+
+            throw bulkWriterStaleError
+        }
+      case e: Throwable =>
+        log.logError(s"Unexpected error when committing writes", e)
+        throw e
+    }
 
     new WriterCommitMessage {}
   }
 
   override def abort(): Unit = {
     log.logInfo("abort invoked!!!")
-    writer.abort()
+    writer.get.abort(true)
     if (cacheItemReleasedCount.incrementAndGet() == 1) {
       clientCacheItem.close()
     }
   }
 
   override def close(): Unit = {
     log.logInfo("close invoked!!!")
-    writer.flushAndClose()
+    writer.get.flushAndClose()
```

Comment:
do we also need to retry here in case any hanging?

Reply:
ACK - Fixed